### PR TITLE
Support Switching The Source Of A Map

### LINKED ISSUE #5391: Map / List: Replace data source using Inspector Data Picker
Success Criteria / demo scenario: 

```
export default function ExampleComponent() {
  const { reviews, recommendedProducts } = useLoaderData()
  return (
    <>
      {[1, 2, 3].map((v) => {
        return <Placeholder />
      })}
    </>
  )
}
```

Select the map in the navigator.
Using the inspector, pick a different data source - `reviews` from `const { reviews } = useLoaderData()` 
The code is now `reviews.map((v) => { return <Placeholder />})`

### PR DESCRIPTION
**Pull request by seanparsons**
**Problem:**
We want to be able to swap the source of a map expression, using their newly refactored representation.

**Fix:**
To support this change, a few bits of logic have been adapted to cater for the more specialised case of updating the source of the map expression. So as to work for both the property of an element case and for the map expression source case.

**Commit Details:**
- Added `variablesInScope` parameter to `addFakeSpyEntry` to be included
  in the metadata.
- Changed `expression` on `UpdateMapExpression` to take a `JSExpression`.
- `DataPickerPopup` now takes a `pickerType` to distinguish between differing
  types of targets for the picker.
- Extracted out `DataPickerPopupSubvariables` component.
- `ValueRow` also now uses `pickerType`.
- Functions like `usePropertyControlDescriptions` now support an optional property.
- Added `ListSection` component to handle the map expressions.
- `CodeElementSection` doesn't show anymore for map expressions.

**Manual Tests:**
I hereby swear that:

- [x] I opened a hydrogen project and it loaded
- [x] I could navigate to various routes in Preview mode

Fixes #5391

### CODE REVIEW

File: editor/karma-base.conf.js [hunk 1]
```diff
@@ -44,6 +44,10 @@ module.exports = function (config) {
     basePath: '',
 
     browserNoActivityTimeout: 1000000,
+    pingTimeout: 15000,
```
[the remark this answers is not in the document]
Comment (seanparsons) [reply editor/karma-base.conf.js]: Fixed in 79da2f3bb.

File: editor/src/components/inspector/code-element-section.spec.browser2.tsx [hunk 2]
```diff
@@ -117,18 +117,12 @@ describe('The Inspector code element section', () => {
...
   })
 
-  it('Is the only section displayed when a map is selected', async () => {
-    const editor = await renderTestEditorWithCode(testProjectCode, 'await-first-dom-report')
-    await clickElementWithTestId(editor, 'ddd')
-
-    // Check that the code element section is displayed
-    const codeElementSection = editor.renderedDOM.getByTestId(CodeElementSectionTestId)
-    expect(codeElementSection).toBeDefined()
+  // it('Is not displayed when a map is selected', async () => {
```
Comment (Rheeseyb) [inline editor/src/components/inspector/code-element-section.spec.browser2.tsx:120]: Does this test still need to be commented out?

Comment (seanparsons) [reply editor/src/components/inspector/code-element-section.spec.browser2.tsx:120]: Probably shouldn't be.

File: editor/src/components/inspector/sections/component-section/variables-in-scope-utils.ts [hunk 3]
```diff
@@ -395,7 +398,11 @@ export function useVariablesInScopeForSelectedElement(
       // if the selected element doesn't have an associacted variablesInScope, we assume it's safe to walk up the path within the same component
       let path = selectedViewPath
       while (
-        EP.pathsEqual(EP.getContainingComponent(path), EP.getContainingComponent(selectedViewPath))
+        EP.pathsEqual(
```
Comment (Rheeseyb) [inline editor/src/components/inspector/sections/component-section/variables-in-scope-utils.ts:401]: I'm like 90% sure we used to have a function which would create all of the ancestor paths of a given element path. Maybe we deleted that at some point, but that sounds much safer than using a `while` loop here

**Review by Rheeseyb — APPROVED** [review]

File: editor/src/components/inspector/code-element-section.spec.browser2.tsx (hunk 2 above)
Comment (seanparsons) [reply editor/src/components/inspector/code-element-section.spec.browser2.tsx]: Restored in 83ae48b86.

File: editor/src/components/inspector/sections/component-section/variables-in-scope-utils.ts (hunk 3 above)
Comment (ruggi) [reply editor/src/components/inspector/sections/component-section/variables-in-scope-utils.ts:401]: what about using `EP.getAncestors` (or possibly `findAmongAncestorsOfPath`) and filtering over those results?

File: editor/src/components/inspector/sections/layout-section/list-section.tsx [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    control: 'array',
+    propertyControl: {
+      control: 'none',
+    },
+  }
+
+  const { popupIsOpen, DataPickerOpener, DataPickerComponent, setReferenceElement, openPopup } =
+    useDataPickerButton(paths)
+
+  // early return, no hooks below this point!
```
Comment (ruggi) [inline editor/src/components/inspector/sections/layout-section/list-section.tsx:199]: This does not seem relevant/correct anymore?

**Review by ruggi — APPROVED** [review]

File: editor/src/components/inspector/sections/component-section/variables-in-scope-utils.ts (hunk 3 above)
Comment (balazsbajorics) [reply editor/src/components/inspector/sections/component-section/variables-in-scope-utils.ts]: fixed in ade133944!

**Pull request merged**